hibernate-reactive-core/src/main/java/org/hibernate/reactive/util/impl/CompletionStages.java

```diff
@@ -222,8 +206,17 @@ public static void logSqlException(Throwable t, Supplier<String> message, String
 	 * }
 	 * </pre>
 	 */
-	public static CompletionStage<Void> loop(int start, int end, Function<Integer,CompletionStage<?>> consumer) {
-		return loop( IntStream.range(start, end), consumer );
+	public static CompletionStage<Void> loop(int start, int end, Function<Integer, CompletionStage<?>> consumer) {
+		if ( start < end ) {
+			return AsyncTrampoline.asyncWhile(
+					index -> index < end,
+					index -> consumer
+							.apply( index )
+							.thenApply( u -> index + 1 ),
+					start
+			).thenCompose( CompletionStages::voidFuture );
+		}
+		return voidFuture();
```

Comment:
Again: I think a much more interesting optimization would be to get rid of the boxing by replacing `Function` with a different SAM.